bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java

```diff
@@ -2579,6 +2639,43 @@ protected void setSelectionToWidget(@SuppressWarnings("rawtypes") List v, boolea
 				}
 			}
 		}
+
+		// there can be some items inside expandable node and not populated yet. In this
+		// case try to find the item to select inside all the visible expandable nodes.
+		if (newSelection.size() < v.size() && getContentProvider() instanceof ILimitBasedContentProvider) {
```

Comment:
**This** code makes Eclipse hang on a trivial "ArrayList" with "limit" of 7 elements in the Outline view (which I've manually updated to play with this PR). We have to solve that differently.